cpp/models/secir/analyze_result.cpp

```diff
@@ -18,52 +18,72 @@
 * limitations under the License.
 */
 #include "secir/analyze_result.h"
+#include "memilio/math/interpolation.h"
 
 #include <algorithm>
 #include <cassert>
 
 namespace mio
 {
 
-/**
- * TODO: extrapolate first and last point
- */
-TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result)
+TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result, const double abs_tol)
 {
-    assert(simulation_result.get_num_time_points() > 0 && "TimeSeries must not be empty.");
-
-    const auto t0      = simulation_result.get_time(0);
-    const auto tmax    = simulation_result.get_last_time();
-    const auto day0    = static_cast<int>(floor(t0));
-    const auto day_max = static_cast<int>(ceil(tmax));
+    const auto t0 = simulation_result.get_time(0);
+    const auto t_max = simulation_result.get_last_time();
+    // add another day if the first time point is equal to day_0 up to absolute tolerance tol
+    const auto day0 = (t0 - abs_tol < std::ceil(t0) - 1) ? std::floor(t0) : std::ceil(t0);
+    // add another day if the last time point is equal to day_max up to absolute tolerance tol
+    const auto day_max = (t_max + abs_tol > std::floor(t_max) + 1) ? std::ceil(t_max) : std::floor(t_max);
+    
+    // create interpolation_times vector with all days between day0 and day_max
+    std::vector<double> tps(static_cast<int>(day_max) - static_cast<int>(day0) + 1);
+    std::iota(tps.begin(), tps.end(), day0);
+    
+    return interpolate_simulation_result(simulation_result, tps, abs_tol);
+}
 
-    auto day = day0;
+TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result, const std::vector<double>& interpolation_times, const double abs_tol)
+{
+    assert(simulation_result.get_num_time_points() > 0 && "TimeSeries must not be empty.");
+    
+    assert((*std::min_element(interpolation_times.begin(), interpolation_times.end()) + abs_tol >= simulation_result.get_time(0) && *std::max_element(interpolation_times.begin(), interpolation_times.end()) - abs_tol <= simulation_result.get_last_time()) && "Interpolation times have lie between simulation times (up to given tolerance).");
+    unused(abs_tol);
```

Comment:
do you really only have the tolerance argument for this assert?

Reply:
Yes, since you pass a vector of the interpolation times, these only have to be checked if the lie between the boundaries of the time series, up to the tolerance. I'm open for suggestions if there is a better solution.

Reply:
I assume the algorithm itself is robust to any tolerance (i.e. doesn't fail or crash), or the argument wouldn't be unused. In that case you don't need to check this condition at all.

If you want to check this, it's ok to not use tolerance at all. It's easy for the user to generate values that lie safely within the (closed) interval. There is no need to use tolerances to check values that you haven't generated yourself.

Your check crashes if `interpolation_times` is empty, by the way.

Reply:
It is correct that the user can easily produce values that lie within the interval.
The only problem I have is that the other function that generates interpolation values for the "days" from a given tolerance might allow for values that lie slightly outside of the interval and then calls the function with this vector that does not lie strictly within the interval. Having a check without tolerance would then lead to an assertion error.
On the other hand, having no check would lead to potentially misleading/wrong results when calling the function with a wrong vector by mistake since no error would be generated.

Reply:
I removed the tolerance in the main analyze_result function and changed the assertions. The function now allows for one time point that is smaller or equal to the first simulation result time and one time point that is larger than the last simulation result time. The values at these time points might not be meaningful and are intended to be used only for small tolerances from the boundaries of the simulation time.